src/uct/ib/dc/verbs/dc_verbs.c

```diff
@@ -0,0 +1,771 @@
+            .ep_atomic_fadd32         = uct_dc_verbs_ep_atomic_fadd32,
+            .ep_atomic_swap32         = uct_dc_verbs_ep_atomic_swap32,
+            .ep_atomic_cswap32        = uct_dc_verbs_ep_atomic_cswap32,
+
+            .ep_flush                 = uct_dc_verbs_ep_flush,
+
+            .ep_pending_add           = uct_dc_ep_pending_add,
+            .ep_pending_purge         = uct_dc_ep_pending_purge,
+        },
+        .arm_tx_cq                = uct_ib_iface_arm_tx_cq,
+        .arm_rx_cq                = uct_ib_iface_arm_rx_cq,
+    },
+    .fc_ctrl                  = NULL /* TODO: */
+};
+
+void uct_dc_verbs_iface_init_wrs(uct_dc_verbs_iface_t *self)
+{
+    
+    /* Initialize inline work request */
+    memset(&self->inl_am_wr, 0, sizeof(self->inl_am_wr));
+    self->inl_am_wr.sg_list                 = self->verbs_common.inl_sge;
+    self->inl_am_wr.num_sge                 = 2;
+    self->inl_am_wr.exp_opcode              = IBV_WR_SEND;
+    self->inl_am_wr.exp_send_flags          = IBV_SEND_INLINE;
+    self->inl_am_wr.dc.dct_access_key       = UCT_IB_DC_KEY;
+
+    memset(&self->inl_rwrite_wr, 0, sizeof(self->inl_rwrite_wr));
+    self->inl_rwrite_wr.sg_list             = self->verbs_common.inl_sge;
+    self->inl_rwrite_wr.num_sge             = 1;
+    self->inl_rwrite_wr.exp_opcode          = IBV_WR_RDMA_WRITE;
+    self->inl_rwrite_wr.exp_send_flags      = IBV_SEND_SIGNALED | IBV_SEND_INLINE;
+    self->inl_rwrite_wr.dc.dct_access_key   = UCT_IB_DC_KEY;
+}
+
+static UCS_CLASS_INIT_FUNC(uct_dc_verbs_iface_t, uct_md_h md, uct_worker_h worker,
+                           const char *dev_name, size_t rx_headroom,
+                           const uct_iface_config_t *tl_config)
+{
+    int i;
+    ucs_status_t status;
+    uct_dc_iface_config_t *config = ucs_derived_of(tl_config,
+                                                   uct_dc_iface_config_t);
+    ucs_trace_func("");
+    UCS_CLASS_CALL_SUPER_INIT(uct_dc_iface_t, &uct_dc_verbs_iface_ops, md,
+                              worker, dev_name, rx_headroom, 0, config);
+
+    self->verbs_common.config.max_inline = config->max_inline;
+
+    status = uct_rc_verbs_iface_common_init(&self->verbs_common, &self->super.super, &config->super);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    uct_dc_verbs_iface_init_wrs(self);
+
+    status = uct_rc_verbs_iface_prepost_recvs_common(&self->super.super);
+    if (status != UCS_OK) {
+        goto out;
+    }
+
+    self->dcis_txcnt = ucs_malloc(self->super.tx.ndci * sizeof(uct_rc_verbs_txcnt_t), "dc");
+    if (self->dcis_txcnt == NULL) {
+        status = UCS_ERR_NO_MEMORY;
+        goto out;
+    }
+
+    for (i = 0; i < self->super.tx.ndci; i++) {
+        uct_rc_verbs_txcnt_init(&self->dcis_txcnt[i]);
+        uct_rc_txqp_available_set(&self->super.tx.dcis[i].txqp, self->super.super.config.tx_qp_len);
+    }
+    
+    uct_worker_progress_register(worker, uct_dc_verbs_iface_progress, self);
```

Comment:
need to register progress only when have an either local/remote connection, because doing progress all the time will hurt performance